src/source-pcap-file-directory-helper.c

```diff
@@ -314,82 +316,107 @@ TmEcode PcapDirectoryPopulateBuffer(PcapFileDirectoryVars *pv,
         SCLogError(SC_ERR_INVALID_ARGUMENT, "No directory filename was passed");
         SCReturnInt(TM_ECODE_FAILED);
     }
+    if ((pv->should_recurse == false) && (pv->cur_dir_depth > 0)) {
+        /* User has not enabled recursion; this is a subdirectory */
+        SCLogInfo("Subdirectory %s will be ignored.  Enable recursive mode using --pcap-file-recursive.", pv->filename);
+        SCReturnInt(TM_ECODE_OK);
+    }
+    if ((pv->should_recurse == true) && (pv->cur_dir_depth >= max_directory_recursion)){
+        /* Maximum allowable depth reached; ending recursive traversal. */
+        SCLogInfo("Directory depth of %d has been reached.  "
+                  "Ignoring subdirectory %s", max_directory_recursion, pv->filename);
+        SCReturnInt(TM_ECODE_OK);
+    }
     struct dirent * dir = NULL;
     PendingFile *file_to_add = NULL;
 
     while ((dir = readdir(pv->directory)) != NULL) {
-#ifndef OS_WIN32
-        if (dir->d_type != DT_REG) {
+        if (!SCIsRegularDirectory(dir) && !SCIsRegularFile(dir)) {
             continue;
         }
-#endif
-        if (strcmp(dir->d_name, ".") == 0 ||
-            strcmp(dir->d_name, "..") == 0) {
+        if (SCIsRegularDirectory(dir)){
+            DIR *cur_dir = pv->directory;
+            char *cur_path = pv->filename;
+
+            /* Create nested directory path */
+            char next_dir [PATH_MAX] = {0};
+            int result = PathJoin(next_dir, PATH_MAX, pv->filename, dir->d_name);
+            if (result != TM_ECODE_OK) {
+                return result;
+            }
+            /* Update pv to point to the nested directory.  Note, this open may fail.
+             * Only set pv->directory on success */
+            DIR *directory = opendir(next_dir);
+            if (directory == NULL) {
+                SCLogWarning(SC_WARN_PATH_READ_ERROR, "Error opening directory %s.  %s", next_dir, strerror(errno));
+                SCReturnInt(TM_ECODE_FAILED);
+            }
+            pv->directory = directory;
+            pv->filename = next_dir;
+            /* Recursively descend into directory structure */
+            ++pv->cur_dir_depth;
+            PcapDirectoryPopulateBuffer(pv, older_than);
+            --pv->cur_dir_depth;
```

Comment:
Can this go below 0? (underflowing)

Guess I'm not following the depth logic yet.

Reply:
There shouldn't be a risk of this wrapping around.  The logic is:
- Recursively descend to the deepest allowable directory to build the directory tree.  This is what `cur_dir_depth` tracks.
- On the return path add each file in the directory to the pending file buffer